stellar_uri.go

```diff
@@ -0,0 +1,336 @@
+var ErrInvalidScheme = errors.New("invalid stellar URI scheme")
+
+// ErrInvalidOperation is returned if the URI operation is not supported.
+var ErrInvalidOperation = errors.New("invalid stellar URI operation")
+
+// ErrBadSignature is returned if the signature fails verification.
+var ErrBadSignature = errors.New("bad signature")
+
+// SignStellarURI signs a stellar+web URI and returns the URI with the signature
+// attached.
+func SignStellarURI(uri string, seed SeedStr) (signedURI, signatureB64 string, err error) {
+	kp, err := keypair.Parse(seed.SecureNoLogString())
+	if err != nil {
+		return "", "", err
+	}
+	payload := payloadFromString(uri)
+	signature, err := kp.Sign(payload)
+	if err != nil {
+		return "", "", err
+	}
+	signatureB64 = base64.StdEncoding.EncodeToString(signature)
+	signatureEsc := url.QueryEscape(signatureB64)
+
+	signedURI = uri + "&signature=" + signatureEsc
+
+	return signedURI, signatureB64, nil
+}
+
+// ValidatedStellarURI contains the origin domain that ValidateStellarURI
+// confirmed
+type ValidatedStellarURI struct {
+	URI          string
+	Operation    string
+	OriginDomain string
+	Message      string
+	CallbackURL  string
+	XDR          string
+	TxEnv        *xdr.TransactionEnvelope
+	Recipient    string
+	Amount       string
+	AssetCode    string
+	AssetIssuer  string
+	Memo         string
+	MemoType     string
+}
+
+// ValidateStellarURI will check the validity of a web+stellar SEP7 URI.
+//
+// It will check that the parameters are valid and that the payload is
+// signed with the appropriate key.
+func ValidateStellarURI(uri string, getter HTTPGetter) (*ValidatedStellarURI, error) {
+	uv, err := newUnvalidatedURI(uri)
+	if err != nil {
+		return nil, err
+	}
+	return uv.Validate(getter)
+}
+
+type unvalidatedURI struct {
+	raw          string
+	values       url.Values
+	Operation    string
+	OriginDomain string
+	Signature    string
+}
+
+func newUnvalidatedURI(uri string) (*unvalidatedURI, error) {
+	res := &unvalidatedURI{raw: uri}
+
+	u, err := url.Parse(uri)
+	if err != nil {
+		return nil, err
+	}
+	if u.Scheme != "web+stellar" {
+		return nil, ErrInvalidScheme
+	}
+
+	res.Operation = u.Opaque
+
+	res.values = u.Query()
+
+	res.OriginDomain = res.value("origin_domain")
+	res.Signature = res.value("signature")
+
+	return res, nil
+
+}
+
+func (u *unvalidatedURI) Validate(getter HTTPGetter) (*ValidatedStellarURI, error) {
+	// origin_domain and signature are optional in the spec, but
+	// it seems like a really bad idea to allow any of these
+	// requests without them, so we are going to make them required.
+	if u.OriginDomain == "" {
+		return nil, ErrMissingParameter{Key: "origin_domain"}
+	}
+	if u.Signature == "" {
+		return nil, ErrMissingParameter{Key: "signature"}
+	}
+
+	// make sure there's no port or scheme
+	if strings.IndexByte(u.OriginDomain, ':') != -1 {
+		return nil, ErrInvalidParameter{Key: "origin_domain"}
+	}
+
+	switch u.Operation {
```

Comment:
Might as well call validateOriginDomain before the switch. So that it doesn't need to be remembered when adding more cases.

Reply:
what if the new operations don't have an origin_domain?

fwiw, the spec doesn't require origin_domain or signature.

Reply:
if the new operation didn't require origin_domain and whoever added the new case forgot about it, then it would fail any tests or real runs where origin_domain wasn't provided